### PR TITLE
fix: improve partial publishing recovery for CocoaPods and GitHub targets

### PR DESCRIPTION
**Pull request by BYK**
## Summary

Fixes partial publishing recovery issues where Craft loses publishing progress and fails unrecoverably on re-run. Triggered by [this CI failure](https://github.com/getsentry/publish/actions/runs/26230754714/job/77190151358).

### CocoaPods: selective retry for `pod trunk push`
- Wraps `pod trunk push` in `withRetry()` with 5 retries and exponential backoff (5s initial, 2x factor)
- Retries only on transient errors (CDN timeouts, 5xx, ETIMEDOUT, ECONNRESET, etc.)
- Fails fast on permanent errors (spec validation, authentication)
- Follows the same pattern as the Crates target

### GitHub: detect existing releases on re-run
- Before `createDraftRelease()`, checks if a release for the tag already exists via new `getReleaseByTag()` method
- If a published release exists (from a previous crashed run), skips creation and returns early
- If a leftover draft release exists, deletes it and creates a fresh one
- Prevents 422 errors on re-run after a partially successful publish

### GitHub: safe rollback with re-fetch
- Before attempting to delete an "orphaned" draft release on failure, re-fetches the release from GitHub via new `getRelease()` method to check its actual state
- If `publishRelease()` half-succeeded (server processed it but response timed out), the release is now published — skips deletion to avoid destroying a live release
- Handles immutable releases gracefully (no crash from failed delete attempts)

### GitHub: non-fatal floating tag updates
- Wraps `updateFloatingTags()` in try/catch after successful `publishRelease()`
- Floating tag failures are logged as warnings instead of throwing
- Prevents marking the target as failed when the release is already live, which would cause duplicate-release errors on re-run

## Tests
- 7 new tests for CocoaPods selective retry (transient vs permanent errors, retry exhaustion)
- 14 new tests for GitHub target (existing release detection, safe rollback, non-fatal floating tags, `getRelease`/`getReleaseByTag` edge cases)

### CODE REVIEW

**Review by BYK — COMMENTED** [review]
Addressed in 2ee4e2a — renamed `COCOAPODS_MAX_RETRIES` to `COCOAPODS_MAX_ATTEMPTS` with updated documentation to match the Crates target convention.

File: src/targets/cocoapods.ts [hunk 1]
```diff
@@ -107,14 +147,45 @@ export class CocoapodsTarget extends BaseTarget {
...
+              );
+              return false;
+            }
+            this.logger.warn(
+              `pod trunk push failed with a transient error, retrying in ${delay}s...`,
+            );
+            this.logger.debug('Error details:', err.message);
+            await sleep(delay * 1000);
+            delay *= COCOAPODS_RETRY_EXP_FACTOR;
+            return true;
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/targets/cocoapods.ts:214]: Valid finding, addressed in 66ab5d2. Added `COCOAPODS_ALREADY_PUBLISHED` constant and a try/catch inside the `withRetry` body (same pattern as the Crates target with `REPUBLISH_ERROR`). When `pod trunk push` fails with "has already been pushed", it's treated as success instead of a permanent failure.

File: src/targets/github.ts [hunk 2]
```diff
@@ -613,6 +667,44 @@ export class GitHubTarget extends BaseTarget {
       ? false
       : isLatestRelease(latestRelease, version);
 
+    // Check for an existing release for this tag. This handles the case
+    // where a previous publish run succeeded in creating/publishing the
+    // release but crashed before persisting state (e.g., floating tag
+    // failure, or a process crash).
+    const tag = versionToTag(version, this.githubConfig.tagPrefix);
+    const existingRelease = await this.getReleaseByTag(tag);
+    if (existingRelease) {
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/targets/github.ts:773]: Not a valid issue. GitHub's `createRelease` with `draft: true` does create the git tag (via `target_commitish`) — draft releases are associated with tags and are returned by `getReleaseByTag`. The GitHub API docs confirm this: "If the tag does not exist, a new tag is created" at release creation time, even for drafts. The `getReleaseByTag` endpoint returns both draft and published releases.

The scenario this code handles is: a previous Craft run created a draft release (which created the tag), then crashed during asset upload. On re-run, `getReleaseByTag` finds the leftover draft, deletes it, and proceeds with a fresh creation.

File: src/targets/github.ts [hunk 3]
```diff
@@ -228,6 +228,86 @@ export class GitHubTarget extends BaseTarget {
...
+   * @param tag The tag name to search for
+   * @returns Array of draft releases matching the tag
+   */
+  public async findDraftReleasesByTag(tag: string): Promise<GitHubRelease[]> {
+    const { data: releases } = await this.github.repos.listReleases({
+      owner: this.githubConfig.owner,
+      repo: this.githubConfig.repo,
+      per_page: 30,
+    });
+    return releases.filter(r => r.tag_name === tag && r.draft === true);
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/targets/github.ts]: Increased `per_page` to 100 (GitHub API max) and added a comment explaining why pagination isn't needed: a leftover draft from a crashed run will always be among the most recent releases. In practice, there can only be one leftover draft per tag, and it was just created in this or a very recent run.

File: src/targets/github.ts [hunk 4]
```diff
@@ -228,6 +228,92 @@ export class GitHubTarget extends BaseTarget {
...
+        try {
+          await this.deleteRelease(draft);
+        } catch (deleteError) {
+          this.logger.warn(
+            `Failed to delete leftover draft release: ${deleteError}`,
+          );
+        }
+      }
+
+      // Retry creation after cleanup
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/targets/github.ts]: Valid point. Fixed in c43a611 — now tracks whether any deletion succeeded. If all deletions fail, re-throws the original 422 instead of blindly retrying `createDraftRelease`.
